crates/ruff_linter/src/rules/ruff/rules/helpers.rs

```diff
@@ -84,3 +87,72 @@ pub(super) fn is_descriptor_class(func: &Expr, semantic: &SemanticModel) -> bool
         })
     })
 }
+
+// fast check to disqualify any string literal that doesn't
+// have opening and closing brackets
+#[inline]
+fn has_brackets(possible_fstring: &str) -> bool {
+    let mut opening = false;
+    let mut closing = false;
+    for c in possible_fstring.chars() {
+        match c {
+            '{' => opening = true,
+            '}' => closing = true,
+            _ => {}
+        }
+    }
+    opening && closing
+}
+
+/// Returns `true` if `source` is valid f-string syntax with qualified, bound variables.
+/// `kwargs` should be the keyword arguments that were passed to function if the string literal is also
+/// being passed to the same function.
+/// If a identifier from `kwargs` is used in `source`'s formatting, this will return `false`,
+/// since it's possible the function could be formatting the literal in question.
+pub(super) fn should_be_fstring(
+    source: &str,
+    kwargs: Option<&Vec<Keyword>>,
+    semantic: &SemanticModel,
+) -> bool {
+    if !has_brackets(source) {
+        return false;
+    }
+
+    let Ok(Expr::FString(ast::ExprFString { value, .. })) =
+        parse_expression(&format!("f\"{source}\""))
+    else {
+        return false;
+    };
+
+    let kw_idents: FxHashSet<String> = kwargs
+        .map(|keywords| {
+            keywords
+                .iter()
+                .filter_map(|k| k.arg.clone())
+                .map(Into::into)
+                .collect()
+        })
+        .unwrap_or_default();
+
+    for f_string in value.f_strings() {
+        let mut has_name = false;
+        for element in &f_string.elements {
+            let Some(ast::FStringExpressionElement { expression, .. }) = element.as_expression()
+            else {
+                continue;
+            };
+
+            if let Expr::Name(ast::ExprName { id, .. }) = expression.as_ref() {
+                if kw_idents.contains(id) || semantic.lookup_symbol(id.as_str()).is_none() {
```

Comment:
Do you mind adding some comments here to guide the reader? E.g.:

Reply:
As a tip, you can use `semantic.resolve_name` instead of `semantic.lookup_symbol`, since you have an `ast::ExprName` here. (We save the binding that each name resolves to, so if you have a name rather than a string symbol, looking up whether its bound is really cheap.)

Reply:
This isn't working for me, and I wonder if it's because the binding id is different (since this `ExprName` came from a standalone `parse_expression` call)

Reply:
Oh yes of course -- my bad. It has the be a node that we visited in the semantic model, and these are entirely new. Sorry!!!